x/interchainquery/keeper/msg_server.go

```diff
@@ -93,6 +94,34 @@ func (k Keeper) VerifyKeyProof(ctx sdk.Context, msg *types.MsgSubmitQueryRespons
 	return nil
 }
 
+// Checks if the query timed out, and if so, checks the timeout policy
+// Returns whether the callback should be executed (only occurs if explicitly indicated)
+func (k Keeper) CheckForQueryTimeout(ctx sdk.Context, query types.Query) (executeCallback bool, err error) {
```

Comment:
Could we either name this more clearly, maybe `CheckForQueryTimeoutAndExecutePolicy`, or pull the policy execution out of this function?

Reply:
This is a control flow function mostly meant to break up the main SubmitQueryResponse function - I don't think it would make sense on the query type. 

I could maybe just move it back into `SubmitQueryResponse` ? It gets a bit sloppy with the branching though

```go
if query.HasTimedOut(ctx) {
	k.Logger(ctx).Error(utils.LogICQCallbackWithHostZone(query.ChainId, query.CallbackId,
		"QUERY TIMEOUT - QueryId: %s, TTL: %d, BlockTime: %d", query.Id, query.TimeoutTimestamp, ctx.BlockHeader().Time.UnixNano()))

	switch query.TimeoutPolicy {
	case types.TimeoutPolicy_REJECT_QUERY_RESPONSE:
		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(query.ChainId, query.CallbackId, "Rejecting query"))
		return nil

	case types.TimeoutPolicy_RETRY_QUERY_REQUEST:
		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(query.ChainId, query.CallbackId, "Retrying query..."))
		return k.RetryICQRequest(ctx, query)

	case types.TimeoutPolicy_EXECUTE_QUERY_CALLBACK:
		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(query.ChainId, query.CallbackId, "Executing callback..."))
	
	        if err := k.InvokeCallback(ctx, msg, query); err != nil {
		       k.Logger(ctx).Error(utils.LogICQCallbackWithHostZone(query.ChainId, query.CallbackId,
			"Error invoking ICQ callback, error: %s, QueryId: %s, QueryType: %s, ConnectionId: %s, QueryRequest: %v, QueryReponse: %v",
			err.Error(), msg.QueryId, query.QueryType, query.ConnectionId, query.RequestData, msg.Result))
		        return  err
                }
                return nil
	}

	default:
		return false, fmt.Errorf("Unsupported query timeout policy: %s", query.TimeoutPolicy.String())
	}
}

if err := k.InvokeCallback(ctx, msg, query); err != nil {
	k.Logger(ctx).Error(utils.LogICQCallbackWithHostZone(query.ChainId, query.CallbackId,
	    "Error invoking ICQ callback, error: %s, QueryId: %s, QueryType: %s, ConnectionId: %s, QueryRequest: %v, QueryReponse: %v",
	    err.Error(), msg.QueryId, query.QueryType, query.ConnectionId, query.RequestData, msg.Result))
	return  err
}
return nil
```

Reply:
(side note, I can move the callback error log to InvokeCallback)

Reply:
we could also take what's above and throw it in a function so then we have

```
if query.HasTimedOut(ctx) {
    k.HandleTImeout(...)
} else {
  k.InvokeCallback()
}
```

I think that's my vote

Reply:
yeah your mental model is right. It's unclear to me whether it's intentionally set up that way to fall into an OOP paradigm or just because it turns out that almost everything needs access to the store.  But I've found it's usually best to just throw everything under the keeper so it's easier to find down the road. 

For instance, our GetValidatorFromHostZone function just takes a list of validators so it doesn't need access to the keeper, and thus is defined as just a standard function,. But I've found myself often doing `k.` and searching for it, thinking I'm getting the name wrong, only to find it's just not a part of the keeper.

I generally use the following framework, and find that I usually run into one of four scenarios:
1. Stateful logic (extends keeper)
2. Business/control flow/main logic (extends keeper)
3. Property like functions on a data structure (extend individual data type in `types`)
4. Stateless utility functions <- this is the one where I often wind up throwing it in the keeper for the reasoning above and because it usually is part of a flow in (2)